Race condition when starting multiple terminals? pyenv: cannot rehash: .pyenv-shim exists

This morning when I logged in after an OS update (MacOS Sonoma 14.1) I got `-bash: python: command not found` in all my restarted terminals.

In one of them, I could also see

```
pyenv: cannot rehash: /Users/fsc-eriker/.pyenv/shims/.pyenv-shim exists
```

Of course, the file no longer existed by the time I checked.

Running `pyenv rehash` in one of the terminals resolved the problem in all of them.

My suspicion is that there is a race condition when many terminals start up in quick succession after a reboot, though I have been running `pyenv` for several years and this is the first time I see this problem.

### Prerequisite
* [X] Make sure your problem is not listed in [the common build problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).
* [X] Make sure no duplicated issue has already been reported in [the pyenv issues](https://github.com/pyenv/pyenv/issues). You should look for closed issues, too.
* [X] Make sure you are not asking us to help solving your specific issue.
  * GitHub issues is opened mainly for development purposes. If you want to ask someone to help solving your problem, go to some community site like [Gitter](https://gitter.im/yyuu/pyenv), [StackOverflow](https://stackoverflow.com/questions/tagged/pyenv), etc.
* [ ] Make sure your problem is not derived from packaging (e.g. [Homebrew](https://brew.sh)).
  * Please refer to the package documentation for the installation issues, etc.
* [ ] Make sure your problem is not derived from plugins.
  * This repository is maintaining `pyenv` and the default `python-build` plugin only. Please refrain from reporting issues of other plugins here.

### Description
- [X] Platform information: MacOS 14.1
- [X] OS architecture (e.g. amd64): Intel
- [X] pyenv version: 2.3.17
- [X] Python version: 3.11.3 (among others)
- [ ] C Compiler information (e.g. gcc 7.3):  clang I guess? I have 15.0.0
- [ ] Please attach the debug trace of the failing command as a gist:
  * Run `env PYENV_DEBUG=1 <faulty command> 2>&1 | tee trace.log` and attach `trace.log`. E.g. if you have a problem with installing Python, run `env PYENV_DEBUG=1 pyenv install -v <version> 2>&1 | tee trace.log` (note the `-v` option to `pyenv install`).
- [ ] If you have a problem with installing Python, please also attach `config.log` from the build directory
  * The build directory is reported after the "BUILD FAILED" message and is usually under `/tmp`.
- [ ] If the build succeeds but the problem is still with the build process (e.g. the resulting Python is missing a feature), please attach
  * the debug trace from reinstalling the faulty version with `env PYENV_DEBUG=1 pyenv install -f -k -v <version> 2>&1 | tee trace.log`
  * `config.log` from the build directory. When using `pyenv install` with `-k` as per above, the build directory will be under `$PYENV_ROOT/sources`.

I have `pyenv-virtualenv` plugin but I imagine it is not involved in this issue. Unfortunately, because there is no way to repro reliably, I cannot rule out that this is caused by Homebrew or the plugin or etc.

Several old issues seem related but too old to be directly relevant as duplicates; #1092, #1145, maybe #1137 and #1747.

--- a/libexec/pyenv-rehash
+++ b/libexec/pyenv-rehash
@@ -19,7 +19,7 @@ acquire_lock() {
   local ret
   set -o noclobber
   # Assuming an old lockfile is stale
-  # Unknowns why this happens for some users but this at least unblocks them
+  # Unknown why this happens for some users but this at least unblocks them
   last_acquire_error="$( { ( echo -n > "$PROTOTYPE_SHIM_PATH"; ) 2>&1 1>&3 3>&1-; } 3>&1)" \
     || { find "$PROTOTYPE_SHIM_PATH" -mmin +10 -exec rm -f {} \; ; ret=1; }
   set +o noclobber

--- a/libexec/pyenv-rehash
+++ b/libexec/pyenv-rehash
@@ -2,7 +2,7 @@
 # Summary: Rehash pyenv shims (run this after installing executables)
 
 set -e
-[ -n "$PYENV_DEBUG" ] && set -x
+[[ -n "$PYENV_DEBUG" ]] && set -x
 
 SHIM_PATH="${PYENV_ROOT}/shims"
 PROTOTYPE_SHIM_PATH="${SHIM_PATH}/.pyenv-shim"
@@ -15,13 +15,15 @@ declare last_acquire_error
 acquire_lock() {
   # Ensure only one instance of pyenv-rehash is running at a time by
   # setting the shell's `noclobber` option and attempting to write to
-  # the prototype shim file. If the file already exists, print a warning
-  # to stderr and exit with a non-zero status.
+  # the prototype shim file.
   local ret
   set -o noclobber
-  last_acquire_error="$( { ( echo -n > "$PROTOTYPE_SHIM_PATH"; ) 2>&1 1>&3 3>&1-; } 3>&1)" || ret=1
+  # Assuming an old lockfile is stale
+  # Unknowns why this happens for some users but this at least unblocks them
+  last_acquire_error="$( { ( echo -n > "$PROTOTYPE_SHIM_PATH"; ) 2>&1 1>&3 3>&1-; } 3>&1)" \
+    || { find "$PROTOTYPE_SHIM_PATH" -mmin +10 -exec rm -f {} \; ; ret=1; }
   set +o noclobber
-  [ -z "${ret}" ]
+  [[ -z "${ret}" ]]
 }
 
 remove_prototype_shim() {

--- a/test/rehash.bats
+++ b/test/rehash.bats
@@ -40,7 +40,7 @@ pyenv: cannot rehash: couldn't acquire lock ${PYENV_ROOT}/shims/.pyenv-shim for 
 
 @test "stale lockfile is removed" {
   mkdir -p "${PYENV_ROOT}/shims"
-  touch -t "$(date --date '-10 min' '+%Y%m%d%H%M.%S' 2>/dev/null || date -v-10M '+%Y%m%d%H%M.%S')" "${PYENV_ROOT}/shims/.pyenv-shim"
+  touch -t 200001010000.00 "${PYENV_ROOT}/shims/.pyenv-shim"
   run pyenv-rehash
   assert_success ""
   assert [ ! -e "${PYENV_ROOT}/shims/.pyenv-shim" ]

--- a/test/rehash.bats
+++ b/test/rehash.bats
@@ -40,6 +40,8 @@ pyenv: cannot rehash: couldn't acquire lock ${PYENV_ROOT}/shims/.pyenv-shim for 
 
 @test "stale lockfile is removed" {
   mkdir -p "${PYENV_ROOT}/shims"
+  #GNU and BSD `date` support generating relative dates via different syntax
+  # but BusyBox `date` used in Bash Docker images doesn't
   touch -t 200001010000.00 "${PYENV_ROOT}/shims/.pyenv-shim"
   run pyenv-rehash
   assert_success ""

--- a/test/rehash.bats
+++ b/test/rehash.bats
@@ -38,6 +38,16 @@ pyenv: cannot rehash: couldn't acquire lock ${PYENV_ROOT}/shims/.pyenv-shim for 
   assert_success
 }
 
+@test "stale lockfile is removed" {
+  mkdir -p "${PYENV_ROOT}/shims"
+  touch -t "$(date --date '-10 min' '+%Y%m%d%H%M.%S' 2>/dev/null || date -v-10M '+%Y%m%d%H%M.%S')" "${PYENV_ROOT}/shims/.pyenv-shim"
+  run pyenv-rehash
+  assert_success ""
+  assert [ ! -e "${PYENV_ROOT}/shims/.pyenv-shim" ]
+}
+
+
+
 @test "creates shims" {
   create_alt_executable_in_version "2.7" "python"
   create_alt_executable_in_version "2.7" "fab"

